Sources/NIO/EventLoop.swift

```diff
@@ -791,3 +763,77 @@ public enum EventLoopError: Error {
     /// Shutting down the `EventLoop` failed.
     case shutdownFailed
 }
+
+fileprivate struct TaskQueue {
+    private let thread = Thread.current
+    private var scheduledTasks = PriorityQueue<ScheduledTask>(ascending: true)
+    private var pendingPushes = [ScheduledTask]()
+    private var pendingRemoves = [ScheduledTask]()
+    private let lock = Lock()
+    private let needCoalesce = Atomic<Bool>(value: false)
+    
+    var isEmpty: Bool {
+        return scheduledTasks.isEmpty
+    }
+    
+    var description: String {
+        return scheduledTasks.description
+    }
+    
+    mutating func push(_ task: ScheduledTask) {
+        if thread.isCurrent {
+            scheduledTasks.push(task)
+            return
+        }
+        lock.lock()
+        needCoalesce.store(true)
+        pendingPushes.append(task)
+        lock.unlock()
+    }
+    
+    func peek() -> ScheduledTask? {
+        return scheduledTasks.peek()
+    }
+    
+    mutating func pop() -> ScheduledTask? {
+        return scheduledTasks.pop()
+    }
+    
+    mutating func remove(_ task: ScheduledTask) {
+        if thread.isCurrent {
+            scheduledTasks.remove(task)
+            return
+        }
+        lock.lock()
+        needCoalesce.store(true)
+        pendingRemoves.append(task)
+        lock.unlock()
+    }
+    
+    mutating func removeReady(_ now: DispatchTime) -> ContiguousArray<() -> Void> {
+        precondition(thread.isCurrent)
+        var readyTasks = ContiguousArray<() -> Void>()
+        while let task = scheduledTasks.peek(), task.readyIn(now) <= .nanoseconds(0) {
+            readyTasks.append(task.task)
+            _ = scheduledTasks.pop()
+        }
+        return readyTasks
+    }
+    
+    mutating func coalesce() {
+        precondition(thread.isCurrent)
```

Comment:
should be `assert` instead of `precondition` as a violation would clearly be a SwiftNIO bug